File: javascript/clock/clock.js
export class Clock {

    #hours;
    #minutes;

    constructor(hours, minutes = 0) {
        if (minutes >= 60) {
            hours += Math.floor(minutes / 60);
        }

        while (minutes < 0) {
            hours--;
            minutes += 60;
        }

        while (hours < 0) {
            hours += 24;
        }

        this.#hours = hours % 24;
        this.#minutes = minutes % 60;
    }

    toString() {
        let h = String(this.#hours).padStart(2, '0');
        let m = String(this.#minutes).padStart(2, '0');
        return `${h}:${m}`
    }

    plus(minutes) {
        return new Clock(this.#hours, this.#minutes + minutes);
    }

    minus(minutes) {
        return new Clock(this.#hours, this.#minutes - minutes);
    }

    equals(otherClock) {
        return this.toString() === otherClock.toString();
    }
}

File: java/clock/src/main/java/Clock.java
class Clock {

    private final static int ONE_HOUR_MINUTES = 60;
    private final static int ONE_DAY_IN_MINUTES = 1440;

    private int minutes;

    Clock(int hours, int minutes) {
        this.minutes = 0;
        add(hours * ONE_HOUR_MINUTES + minutes);
    }

    void add(int minutes) {
        int m = (this.minutes + minutes) % ONE_DAY_IN_MINUTES;
        this.minutes = m >= 0 ? m : m + ONE_DAY_IN_MINUTES;
    }

    @Override
    public String toString() {
        return String.format("%02d:%02d", this.minutes / ONE_HOUR_MINUTES, this.minutes % ONE_HOUR_MINUTES);
    }

    @Override
    public boolean equals(Object obj) {
        if (this.getClass() != obj.getClass()) {
            return false;
        }

        return this.toString().equals(obj.toString());
    }
}